File: axon-bridge-vscode/src/BridgeLogService.ts
/**
 * BridgeLogService.ts
 *
 * VS Code-side mirror of Axon's BridgeLogService.
 * Captures incoming/outgoing WebSocket JSON-RPC traffic, validates structure,
 * pretty-prints, retains a bounded buffer, and mirrors output to an OutputChannel.
 */

import * as vscode from 'vscode';

export type BridgeLogDirection = 'incoming' | 'outgoing';
export type BridgeLogMessageType = 'request' | 'response' | 'notification' | 'error' | 'unknown';

export interface BridgeLogEntry {
    id: string;
    timestamp: number; // epoch ms
    direction: BridgeLogDirection;
    messageType: BridgeLogMessageType;
    method?: string;
    requestId?: string;
    rawJSON: string;
    prettyJSON: string;
    isValid: boolean;
    validationErrors: string[];
}

export interface BridgeLogFilter {
    filterText: string;
    showIncoming: boolean;
    showOutgoing: boolean;
    showRequests: boolean;
    showResponses: boolean;
    showNotifications: boolean;
    showErrors: boolean;
    onlyShowInvalid: boolean;
}

const DEFAULT_FILTER: BridgeLogFilter = {
    filterText: '',
    showIncoming: true,
    showOutgoing: true,
    showRequests: true,
    showResponses: true,
    showNotifications: true,
    showErrors: true,
    onlyShowInvalid: false,
};

export class BridgeLogService {
    private static _shared: BridgeLogService | undefined;

    static get shared(): BridgeLogService {
        if (!this._shared) {
            this._shared = new BridgeLogService();
        }
        return this._shared;
    }

    private entries: BridgeLogEntry[] = [];
    private maxEntries = 500;

    private output: vscode.OutputChannel;

    private constructor() {
        this.output = vscode.window.createOutputChannel('Axon Bridge');
    }

    showOutput() {
        this.output.show(true);
    }

    clear() {
        this.entries = [];
        this.output.clear();
    }

    getEntries(): BridgeLogEntry[] {
        return this.entries;
    }

    getFilteredEntries(filter: BridgeLogFilter = DEFAULT_FILTER): BridgeLogEntry[] {
        const text = filter.filterText.trim().toLowerCase();

        return this.entries.filter(e => {
            if (e.direction === 'incoming' && !filter.showIncoming) return false;
            if (e.direction === 'outgoing' && !filter.showOutgoing) return false;

            switch (e.messageType) {
                case 'request':
                    if (!filter.showRequests) return false;
                    break;
                case 'response':
                    if (!filter.showResponses) return false;
                    break;
                case 'notification':
                    if (!filter.showNotifications) return false;
                    break;
                case 'error':
                    if (!filter.showErrors) return false;
                    break;
                case 'unknown':
                    break;
            }

            if (filter.onlyShowInvalid && e.isValid) return false;

            if (text) {
                const matchesMethod = (e.method ?? '').toLowerCase().includes(text);
                const matchesId = (e.requestId ?? '').toLowerCase().includes(text);
                const matchesJson = e.rawJSON.toLowerCase().includes(text);
                if (!matchesMethod && !matchesId && !matchesJson) return false;
            }

            return true;
        });
    }

    export(): string {
        return JSON.stringify(this.entries, null, 2);
    }

    logIncoming(raw: string) {
        this.addEntry(this.createEntry('incoming', raw));
    }

    logOutgoing(raw: string) {
        this.addEntry(this.createEntry('outgoing', raw));
    }

    private addEntry(entry: BridgeLogEntry) {
        // Newest first
        this.entries.unshift(entry);
        if (this.entries.length > this.maxEntries) {
            this.entries = this.entries.slice(0, this.maxEntries);
        }

        // Mirror to OutputChannel
        this.output.appendLine(this.formatForOutput(entry));
    }

    private formatForOutput(entry: BridgeLogEntry): string {
        const ts = new Date(entry.timestamp);
        const time = ts.toISOString().split('T')[1]?.replace('Z', '') ?? ts.toISOString();

        const dir = entry.direction === 'outgoing' ? '→' : '←';
        const type = entry.messageType.toUpperCase();
        const summary = entry.method
            ? entry.method
            : entry.messageType === 'response'
                ? `response ${entry.requestId ?? ''}`
                : entry.messageType;

        const validity = entry.isValid ? '' : ` INVALID(${entry.validationErrors.join('; ')})`;

        return `[${time}] ${dir} ${type} ${summary}${validity}\n${entry.prettyJSON}\n`;
    }

    private createEntry(direction: BridgeLogDirection, raw: string): BridgeLogEntry {
        const timestamp = Date.now();
        const id = `${timestamp}-${Math.random().toString(16).slice(2)}`;

        let messageType: BridgeLogMessageType = 'unknown';
        let method: string | undefined;
        let requestId: string | undefined;

        let prettyJSON = raw;
        let isValid = true;
        let validationErrors: string[] = [];

        try {
            const obj = JSON.parse(raw) as any;

            method = typeof obj.method === 'string' ? obj.method : undefined;
            requestId = obj.id !== undefined ? String(obj.id) : undefined;

            if (obj.error !== undefined) {
                messageType = 'error';
            } else if (obj.result !== undefined) {
                messageType = 'response';
            } else if (obj.method !== undefined) {
                messageType = obj.id !== undefined ? 'request' : 'notification';
            }

            try {
                prettyJSON = JSON.stringify(obj, Object.keys(obj).sort(), 2);
            } catch {
                // ignore
            }

            validationErrors = this.validateJSONRPC(obj);
            if (validationErrors.length > 0) {
                isValid = false;
            }
        } catch (e) {
            isValid = false;
            validationErrors = ['Invalid JSON: Unable to parse'];
        }

        return {
            id,
            timestamp,
            direction,
            messageType,
            method,
            requestId,
            rawJSON: raw,
            prettyJSON,
            isValid,
            validationErrors,
        };
    }

    private validateJSONRPC(obj: any): string[] {
        const errors: string[] = [];

        if (obj?.jsonrpc !== '2.0') {
            if (obj?.jsonrpc === undefined) {
                errors.push('Missing required field: jsonrpc');
            } else {
                errors.push(`jsonrpc field should be "2.0", got "${String(obj.jsonrpc)}"`);
            }
        }

        const hasMethod = obj?.method !== undefined;
        const hasResult = obj?.result !== undefined;
        const hasError = obj?.error !== undefined;

        if (hasResult || hasError) {
            if (obj?.id === undefined) {
                errors.push('Response missing required field: id');
            }
            if (hasResult && hasError) {
                errors.push('Response cannot have both result and error');
            }
        } else if (hasMethod) {
            if (typeof obj.method !== 'string') {
                errors.push('Method must be a string');
            } else if (!obj.method) {
                errors.push('Method cannot be empty');
            }
        } else {
            errors.push('Message must have method (request/notification) or result/error (response)');
        }

        if (obj?.error !== undefined) {
            if (typeof obj.error !== 'object' || obj.error === null) {
                errors.push('Error must be an object');
            } else {
                if (obj.error.code === undefined) errors.push('Error missing required field: code');
                else if (!Number.isInteger(obj.error.code)) errors.push('Error code must be an integer');

                if (obj.error.message === undefined) errors.push('Error missing required field: message');
                else if (typeof obj.error.message !== 'string') errors.push('Error message must be a string');
            }
        }

        return errors;
    }
}

File: Axon/Services/Bridge/BridgeLogService.swift
//
//  BridgeLogService.swift
//  Axon
//
//  Service for capturing and storing VS Code Bridge WebSocket traffic
//  for debugging and inspection purposes.
//

import Foundation
import Combine

// MARK: - Log Entry Model

/// A single log entry representing a WebSocket message
struct BridgeLogEntry: Identifiable, Equatable, Hashable {
    let id: UUID
    let timestamp: Date
    let direction: Direction
    let messageType: MessageType
    let method: String?
    let requestId: String?
    let rawJSON: String
    let prettyJSON: String
    let isValid: Bool
    let validationErrors: [String]

    enum Direction: String, Codable {
        case outgoing = "→"  // Axon to VS Code
        case incoming = "←"  // VS Code to Axon

        var icon: String {
            switch self {
            case .outgoing: return "arrow.up.circle"
            case .incoming: return "arrow.down.circle"
            }
        }

        var label: String {
            switch self {
            case .outgoing: return "Sent"
            case .incoming: return "Received"
            }
        }
    }

    enum MessageType: String, Codable {
        case request = "Request"
        case response = "Response"
        case notification = "Notification"
        case error = "Error"
        case unknown = "Unknown"

        var icon: String {
            switch self {
            case .request: return "arrow.right.circle"
            case .response: return "arrow.left.circle"
            case .notification: return "bell"
            case .error: return "exclamationmark.triangle"
            case .unknown: return "questionmark.circle"
            }
        }
    }

    init(
        direction: Direction,
        rawData: Data,
        timestamp: Date = Date()
    ) {
        self.id = UUID()
        self.timestamp = timestamp
        self.direction = direction
        self.rawJSON = String(data: rawData, encoding: .utf8) ?? "(binary data)"

        // Try to parse and validate the JSON
        var parsedMethod: String? = nil
        var parsedRequestId: String? = nil
        var parsedMessageType: MessageType = .unknown
        var jsonValid = true
        var errors: [String] = []
        var pretty = self.rawJSON

        if let json = try? JSONSerialization.jsonObject(with: rawData) as? [String: Any] {
            // Pretty print
            if let prettyData = try? JSONSerialization.data(withJSONObject: json, options: [.prettyPrinted, .sortedKeys]),
               let prettyString = String(data: prettyData, encoding: .utf8) {
                pretty = prettyString
            }

            // Extract fields
            parsedMethod = json["method"] as? String
            parsedRequestId = json["id"] as? String

            // Determine message type
            if json["error"] != nil {
                parsedMessageType = .error
            } else if json["result"] != nil {
                parsedMessageType = .response
            } else if json["method"] != nil {
                if json["id"] != nil {
                    parsedMessageType = .request
                } else {
                    parsedMessageType = .notification
                }
            }

            // Validate JSON-RPC 2.0 structure
            errors = BridgeLogEntry.validateJSONRPC(json)
            if !errors.isEmpty {
                jsonValid = false
            }
        } else {
            jsonValid = false
            errors = ["Invalid JSON: Unable to parse"]
        }

        self.prettyJSON = pretty
        self.method = parsedMethod
        self.requestId = parsedRequestId
        self.messageType = parsedMessageType
        self.isValid = jsonValid
        self.validationErrors = errors
    }

    /// Validate that the JSON conforms to JSON-RPC 2.0 spec
    private static func validateJSONRPC(_ json: [String: Any]) -> [String] {
        var errors: [String] = []

        // Must have jsonrpc field
        if let jsonrpc = json["jsonrpc"] as? String {
            if jsonrpc != "2.0" {
                errors.append("jsonrpc field should be \"2.0\", got \"\(jsonrpc)\"")
            }
        } else {
            errors.append("Missing required field: jsonrpc")
        }

        // Requests and Responses need id
        let hasMethod = json["method"] != nil
        let hasResult = json["result"] != nil
        let hasError = json["error"] != nil

        if hasResult || hasError {
            // This is a response
            if json["id"] == nil {
                errors.append("Response missing required field: id")
            }
            if hasResult && hasError {
                errors.append("Response cannot have both result and error")
            }
        } else if hasMethod {
            // This is a request or notification
            if let method = json["method"] as? String {
                if method.isEmpty {
                    errors.append("Method cannot be empty")
                }
            } else {
                errors.append("Method must be a string")
            }
        } else {
            errors.append("Message must have method (request/notification) or result/error (response)")
        }

        // Validate error structure if present
        if let error = json["error"] as? [String: Any] {
            if error["code"] == nil {
                errors.append("Error missing required field: code")
            } else if !(error["code"] is Int) {
                errors.append("Error code must be an integer")
            }
            if error["message"] == nil {
                errors.append("Error missing required field: message")
            } else if !(error["message"] is String) {
                errors.append("Error message must be a string")
            }
        }

        return errors
    }

    var formattedTimestamp: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss.SSS"
        return formatter.string(from: timestamp)
    }

    var summary: String {
        if let method = method {
            return method
        } else if messageType == .response {
            return "Response \(requestId ?? "")"
        } else if messageType == .error {
            return "Error"
        }
        return messageType.rawValue
    }
}

// MARK: - Log Service

@MainActor
class BridgeLogService: ObservableObject {
    static let shared = BridgeLogService()

    // MARK: - Published State

    @Published private(set) var entries: [BridgeLogEntry] = []
    @Published var isLoggingEnabled: Bool = true
    @Published var maxEntries: Int = 500

    // MARK: - Filtering

    @Published var filterText: String = ""
    @Published var showIncoming: Bool = true
    @Published var showOutgoing: Bool = true
    @Published var showRequests: Bool = true
    @Published var showResponses: Bool = true
    @Published var showNotifications: Bool = true
    @Published var showErrors: Bool = true
    @Published var onlyShowInvalid: Bool = false

    var filteredEntries: [BridgeLogEntry] {
        entries.filter { entry in
            // Direction filter
            if entry.direction == .incoming && !showIncoming { return false }
            if entry.direction == .outgoing && !showOutgoing { return false }

            // Message type filter
            switch entry.messageType {
            case .request: if !showRequests { return false }
            case .response: if !showResponses { return false }
            case .notification: if !showNotifications { return false }
            case .error: if !showErrors { return false }
            case .unknown: break
            }

            // Validity filter
            if onlyShowInvalid && entry.isValid { return false }

            // Text filter
            if !filterText.isEmpty {
                let searchText = filterText.lowercased()
                let matchesMethod = entry.method?.lowercased().contains(searchText) ?? false
                let matchesJSON = entry.rawJSON.lowercased().contains(searchText)
                let matchesId = entry.requestId?.lowercased().contains(searchText) ?? false
                if !matchesMethod && !matchesJSON && !matchesId { return false }
            }

            return true
        }
    }

    private init() {}

    // MARK: - Logging Methods

    /// Log an outgoing message (Axon → VS Code)
    func logOutgoing(_ data: Data) {
        guard isLoggingEnabled else { return }

        let entry = BridgeLogEntry(direction: .outgoing, rawData: data)
        addEntry(entry)
    }

    /// Log an incoming message (VS Code → Axon)
    func logIncoming(_ data: Data) {
        guard isLoggingEnabled else { return }

        let entry = BridgeLogEntry(direction: .incoming, rawData: data)
        addEntry(entry)
    }

    private func addEntry(_ entry: BridgeLogEntry) {
        entries.insert(entry, at: 0)

        // Trim old entries if over limit
        if entries.count > maxEntries {
            entries = Array(entries.prefix(maxEntries))
        }
    }

    // MARK: - Management

    func clear() {
        entries.removeAll()
    }

    func export() -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601

        let exportData = entries.map { entry -> [String: Any] in
            return [
                "timestamp": ISO8601DateFormatter().string(from: entry.timestamp),
                "direction": entry.direction.rawValue,
                "messageType": entry.messageType.rawValue,
                "method": entry.method as Any,
                "requestId": entry.requestId as Any,
                "isValid": entry.isValid,
                "validationErrors": entry.validationErrors,
                "json": entry.rawJSON
            ]
        }

        if let data = try? JSONSerialization.data(withJSONObject: exportData, options: [.prettyPrinted]),
           let string = String(data: data, encoding: .utf8) {
            return string
        }

        return "[]"
    }

    // MARK: - Statistics

    var requestCount: Int {
        entries.filter { $0.messageType == .request }.count
    }

    var responseCount: Int {
        entries.filter { $0.messageType == .response }.count
    }

    var errorCount: Int {
        entries.filter { $0.messageType == .error || !$0.isValid }.count
    }

    var invalidCount: Int {
        entries.filter { !$0.isValid }.count
    }
}
